### PR TITLE
Add SSVM_Process mode and ssvm_process host modules.

### PR DESCRIPTION
**Pull request by q82419**

### CODE REVIEW

**Review by hydai — CHANGES_REQUESTED** [review]

File: lib/host/ssvm_process/processfunc.cpp [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+                                     uint32_t ArgPtr, uint32_t ArgLen) {
+  /// Check memory instance from module.
+  if (MemInst == nullptr) {
+    return Unexpect(ErrCode::ExecutionFailed);
+  }
+
+  char *Buf = MemInst->getPointer<char *>(ArgPtr);
+  std::string NewArg;
+  std::copy_n(Buf, ArgLen, std::back_inserter(NewArg));
+  Env.Args.push_back(std::move(NewArg));
```
[the remark this answers is not in the document]
Comment (q82419) [reply lib/host/ssvm_process/processfunc.cpp:38]: Sure.

File: include/vm/costtable.h [hunk 2]
```diff
@@ -27,12 +27,9 @@ class CostTable {
   bool setCostTable(const Configure::VMType &Type) {
     switch (Type) {
     case Configure::VMType::Wasm:
-      /// Wasm cost table
```
[the remark this answers is not in the document]
Comment (q82419) [reply include/vm/costtable.h:30]: In the current state, we don't need a cost table. This is preserved for future using.

File: include/vm/costtable.h [hunk 3]
```diff
@@ -27,12 +27,9 @@ class CostTable {
...
-      Costs[Type] = std::vector<uint64_t>(256, 1);
-      return true;
     case Configure::VMType::Wasi:
-      /// Wasi cost table
```
[the remark this answers is not in the document]
Comment (q82419) [reply include/vm/costtable.h:34]: In the current state, we don't need a cost table. This is preserved for future using.

File: include/vm/costtable.h [hunk 4]
```diff
@@ -43,7 +40,7 @@ class CostTable {
   /// Set customized cost table.
   bool setCostTable(const Configure::VMType &Type,
                     const std::vector<uint64_t> &Table) {
-    if (Table.size() < 0xFF) {
+    if (Table.size() < UINT16_MAX) {
```
[the remark this answers is not in the document]
Comment (q82419) [reply include/vm/costtable.h:43]: I think this work can be done when refactoring in the future.

File: lib/vm/vm.cpp [hunk 5]
```diff
@@ -24,14 +25,23 @@ void VM::initVM() {
...
     Measure.setCostTable(CostTab.getCostTable(Configure::VMType::Wasi));
   }
+  if (Config.hasVMType(Configure::VMType::SSVM_Process)) {
+    /// 1st priority of cost table: SSVM_Process
+    std::unique_ptr<Runtime::ImportObject> ProcMod =
+        std::make_unique<Host::SSVMProcessModule>();
+    InterpreterEngine.registerModule(StoreRef, *ProcMod.get());
+    ImpObjs.insert({Configure::VMType::SSVM_Process, std::move(ProcMod)});
+    CostTab.setCostTable(Configure::VMType::SSVM_Process);
+    Measure.setCostTable(CostTab.getCostTable(Configure::VMType::SSVM_Process));
```
[the remark this answers is not in the document]
Comment (q82419) [reply lib/vm/vm.cpp:43]: The highest priority of the cost table will be set into measurement class.

File: lib/vm/vm.cpp [hunk 6]
```diff
@@ -24,14 +25,23 @@ void VM::initVM() {
   CostTab.setCostTable(Configure::VMType::Wasm);
   Measure.setCostTable(CostTab.getCostTable(Configure::VMType::Wasm));
   if (Config.hasVMType(Configure::VMType::Wasi)) {
-    /// 2nd priority of cost table: Wasi
+    /// 2st priority of cost table: Wasi
```
[the remark this answers is not in the document]
Comment (q82419) [reply lib/vm/vm.cpp:28]: The ideal priority will be:
 1. SSVM_Process
 2. wasi
 3. wasm

File: include/vm/costtable.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (hydai) [reply include/vm/costtable.h:30]: If we don't initialize the cost table, what will happen when we try to get the runtime cost?

File: include/vm/costtable.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (hydai) [reply include/vm/costtable.h:34]: Ditto.

File: include/vm/costtable.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (hydai) [reply include/vm/costtable.h:43]: Okay.

File: lib/host/ssvm_process/processfunc.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (hydai) [reply lib/host/ssvm_process/processfunc.cpp:38]: How about using `AddArg` or `AppendArg`?

File: lib/vm/vm.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (hydai) [reply lib/vm/vm.cpp:43]: Why does the 1st priority be initiated after the 2nd priority?

**Review by hydai — CHANGES_REQUESTED** [review]

File: lib/vm/vm.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (hydai) [reply lib/vm/vm.cpp]: Okay

File: lib/vm/vm.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (q82419) [reply lib/vm/vm.cpp:43]: If 1st priority is initiated before the 2nd priority, the cost table will be overridden when satisfying with the condition of the 2nd priority.

[the remark this answers is not in the document]
Comment (hydai) [reply lib/vm/vm.cpp:43]: If I enable multiple VMType (Wasi+SSVM_Process) at the same time, will the cost calculate correctly?
You said that the cost table will be overridden.
Assume I have:
1) Instruction A in Wasi, cost 100
2) Instruction B in SSVM_Process, cost 200

When I call A, the final cost will be 100.
When I call A+B, what's the final cost? Can it found the instruction cost after its cost table is overridden?

File: include/vm/costtable.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (q82419) [reply include/vm/costtable.h:30]: In this case, the cost table of `VMType::Wasm` will be returned.
I've updated to add default `Wasm` cost table in the constructor.

File: include/vm/costtable.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (q82419) [reply include/vm/costtable.h:34]: Ditto.

File: lib/host/ssvm_process/processfunc.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (q82419) [reply lib/host/ssvm_process/processfunc.cpp:38]: Updated.

File: lib/vm/vm.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (q82419) [reply lib/vm/vm.cpp:43]: The final cost will be 200.
The cost table priority will be `SSVM_Process` > `Wasi` > `Wasm`.

[the remark this answers is not in the document]
Comment (hydai) [reply lib/vm/vm.cpp:43]: Ah-ha, I got it.
The cost table only contains the wasm instructions from wasm spec, not including the host functions part.

**Review by hydai — APPROVED** [review]

**Pull request merged**